### PR TITLE
Add Regex named capture groups

### PR DESCRIPTION
**Pull request by juancarlospaco**
- Added Named Capture Groups, get the captured matches by name.
- https://caniuse.com/mdn-javascript_builtins_regexp_named_capture_groups
- `runnableExamples` with `assert`, parsing an ISO-8601 date/time string. Tiny diff. Good performance.
- Reduces Index error bugs in the Nim code, because theres no indexing.

JavaScript:
```javascript
var [a, b, c] = regex.exec(pattern).slice(1);
```

Nim:
```nim
regex.groups(pattern, a, b, c)
```

### CODE REVIEW

File: lib/js/jsre.nim [hunk 1]
```diff
@@ -51,6 +51,38 @@ func contains*(pattern: cstring; self: RegExp): bool =
     assert "xabc".contains jsregex
   asm "`result` = `self`.test(`pattern`);"
 
+template groups*(self: RegExp; pattern: cstring; groups: varargs[var cstring]) =
```
Comment (xflywind) [inline lib/js/jsre.nim:54]: > `varargs[var cstring]`

why?

**Comment by juancarlospaco** [comment]
@xflywind Can be `openArray` if you want too, but I think `vararg` is Ok.

File: lib/js/jsre.nim (hunk 1 above)
Comment (Araq) [inline lib/js/jsre.nim:54]: Interface is too weird.

File: lib/js/jsre.nim [hunk 2]
```diff
@@ -51,6 +51,38 @@ func contains*(pattern: cstring; self: RegExp): bool =
...
+
+    block:
+      let rex = newRegExp(isoRe)
+      var year, month, day, hour, minute, second, offByOne: cstring
+      # "offByOne" is an extra argument, no bug, no index error.
+      rex.groups "2021-02-31T12:59:30.666", year, month, day, hour, minute, second, offByOne
+      assert offByOne == cstring.default
+
+  doAssert groups.len > 0, "groups must not be empty varargs[var cstring]"
+  {.emit: [groups, " = ", self, ".exec('", pattern, "').slice(1);"] .}
```
Comment (Araq) [inline lib/js/jsre.nim:84]: Hacky implementation.

File: lib/js/jsre.nim (hunk 1 above)
Comment (Araq) [reply lib/js/jsre.nim:54]: `pattern` is **not** the pattern! That would be `self`...

**Pull request closed without merging**